### PR TITLE
Arm64: Support additional condition checks in select nodes

### PR DESCRIPTION
**Pull request by a74nh**
Generating a FEQ or FNEU can generate an incorrect compare.

Firstly, the optimizer needs changing to ensure the NAN flag is correct.

Once that is fixed, it's not possible to generate a test to trigger the code in codegen (as there are no CIL beq.un or bne instructions). However, the codegen code was wrong.

Alternatively, given it's untested, it might be better to replace the codegen changes with some asserts?

I've added some general compare consume tests to ensure everything is generated ok.

### CODE REVIEW

**Comment by jakobbotsch** [comment]
> Once that is fixed, it's not possible to generate a test to trigger the code in codegen (as there are no CIL beq.un or bne instructions). However, the codegen code was wrong.

I assume it can happen due to morph calling `gtReverseCond` itself in a bunch of places?

I think I've commented on this before, but I really would prefer if we just swapped the operands in the select node we create below (which seems more obviously correct to me, without having to rely on `gtReverseCond` not hitting the `GT_NOT` case).

**Review by jakobbotsch — APPROVED** [review]

**Comment by a74nh** [comment]
>I assume it can happen due to morph calling gtReverseCond itself in a bunch of places?

It's not quite that simple - you can't reverse a "single op" compare reverse into a non single op compare.

Ordered Float NE -> Unordered Float EQ -> Ordered Float NE -> etc (always multi op)
Unordered Float NE -> Ordered Float EQ -> Unordered Float NE -> etc (always single op)

**Comment by jakobbotsch** [comment]
> It's not quite that simple - you can't reverse a "single op" compare reverse into a non single op compare.

Gotcha, makes sense. So essentially we never see these multi-ops on ARM64.

> Alternatively, given it's untested, it might be better to replace the codegen changes with some asserts?

I will leave it up to you. The handling you added in this PR in the backend looks innocent/correct enough.

**Comment by a74nh** [comment]
> I will leave it up to you. The handling you added in this PR in the backend looks innocent/correct enough.

Ok, will leave as is then.

**Comment by jakobbotsch** [comment]
Failures are #78290 (the CI run started right before the fix was merged)

**Comment by a74nh** [comment]
> Failures are #78290 (the CI run started right before the fix was merged)

Are you sure? This PR isn't merged yet.

**Comment by jakobbotsch** [comment]
I mean the CI failures in this PR are #78290 (which is fixed, but the fix for that issue didn't make it to this PR's CI run).

**Comment by a74nh** [comment]
> I mean the CI failures in this PR are #78290 (which is fixed, but the fix for that issue didn't make it to this PR's CI run).

Ah, sorry, Understood.

**Pull request merged**
